test/test_linalg.py

```diff
@@ -921,6 +922,117 @@ def test_nuclear_norm_exceptions_old(self, device):
+            if 'cuda' in device and dtype.is_complex:
+                result_identity_list1 = []
+                result_identity_list2 = []
+                p = int(np.prod(batches))  # use `p` instead of -1, so that the test works for empty input as well
+                for m, m_inv in zip(matrix.contiguous().view(p, n, n), matrix_inverse.contiguous().view(p, n, n)):
+                    result_identity_list1.append(torch.matmul(m, m_inv))
+                    result_identity_list2.append(torch.matmul(m_inv, m))
+                result_identity1 = torch.stack(result_identity_list1).view(*batches, n, n)
+                result_identity2 = torch.stack(result_identity_list2).view(*batches, n, n)
+                self.assertEqual(identity.expand_as(matrix), result_identity1)
+                self.assertEqual(identity.expand_as(matrix), result_identity2)
+            else:
+                self.assertEqual(identity.expand_as(matrix), torch.matmul(matrix, matrix_inverse))
+                self.assertEqual(identity.expand_as(matrix), torch.matmul(matrix_inverse, matrix))
+
+            # check the out= variant
+            matrix_inverse_out = torch.empty(*batches, n, n, dtype=dtype, device=device)
+            ans = torch.inverse(matrix, out=matrix_inverse_out)
+            self.assertEqual(matrix_inverse_out, ans, atol=0, rtol=0)
+            self.assertEqual(matrix_inverse_out, matrix_inverse, atol=0, rtol=0)
+
+            # batched matrices: 3+ dimensional tensors, check matrix_inverse same as single-inverse for each matrix
+            if matrix.ndim > 2:
+                expected_inv_list = []
+                p = int(np.prod(batches))  # use `p` instead of -1, so that the test works for empty input as well
+                for mat in matrix.contiguous().view(p, n, n):
+                    expected_inv_list.append(torch.inverse(mat))
+                expected_inv = torch.stack(expected_inv_list).view(*batches, n, n)
+                if 'cuda' in device and dtype in [torch.float32, torch.complex64]:
+                    # single-inverse is done using cuSOLVER, while batched inverse is done using MAGMA
+                    # individual values can be significantly different for fp32, hence rather high rtol is used
+                    # the important thing is that torch.inverse passes above checks with identity
+                    self.assertEqual(matrix_inverse, expected_inv, atol=1e-1, rtol=1e-2)
+                else:
+                    self.assertEqual(matrix_inverse, expected_inv)
+
+        for batches, n in itertools.product(
+            [[], [1], [4], [2, 3]],
+            [0, 5, 64]
+        ):
+            # large batch size and large matrix size will be tested in test_inverse_many_batches (slow test)
+            if batches and batches[0] == 32 and n == 256:
+                continue
+            matrices = random_fullrank_matrix_distinct_singular_value(n, *batches, dtype=dtype).to(device)
```

Comment:
why do you need `.to(device)` here instead of using `device` kwarg?

Reply:
But then it's used in the kwargs in method_tests below? (see line 1463 in common_method_invocations below). Also, why is svd so extremely slow on the gpu, and why don't we copy data to the cpu and do it on cpu if that's the case?

Reply:
I think `device` kwarg is used in some old test code when people didn't realize that. We can modify them if needed. 

About why svd is so slow on GPU, hmm, maybe some MAGMA implementation thing? The next step is to add cusolver svd implementation if it performs better.